src/notice/notice_backtrace.rs

```diff
@@ -0,0 +1,141 @@
+use crate::backtrace::{Backtrace, BacktraceFrame, BacktraceSymbol};
+use serde::ser::{Serialize, SerializeSeq, Serializer};
+use serde_json::{self, Value};
+use std::collections::HashMap;
+
+#[derive(Debug, Clone)]
+pub struct NoticeTrace {
+    frames: Vec<NoticeFrame>,
+}
+
+impl NoticeTrace {
+    fn new(frames: Vec<NoticeFrame>) -> NoticeTrace {
+        NoticeTrace { frames }
+    }
+
+    pub fn frames(&self) -> Vec<NoticeFrame> {
+        self.frames.clone()
+    }
+}
+
+impl From<&Backtrace> for NoticeTrace {
+    fn from(backtrace: &Backtrace) -> NoticeTrace {
+        let mut frames: Vec<NoticeFrame> = vec![];
+        for f in backtrace.frames() {
+            frames.append(&mut NoticeFrame::unroll_frame_symbols(&f));
+        }
+        NoticeTrace { frames }
+    }
+}
+
+impl Serialize for NoticeTrace {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.frames.len()))?;
+        for element in self.frames() {
+            seq.serialize_element(&element)?;
+        }
+        seq.end()
+    }
+}
+
+impl From<NoticeTrace> for Value {
+    fn from(notice_backtrace: NoticeTrace) -> Value {
+        serde_json::json!(notice_backtrace)
+    }
+}
+
+#[derive(Debug, Serialize, Clone)]
+pub struct NoticeFrame {
+    pub file: String,
+    pub function: String,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub line: Option<u32>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub code: Option<HashMap<i32, String>>,
+}
+
+impl NoticeFrame {
+    fn unroll_frame_symbols(frame: &BacktraceFrame) -> Vec<NoticeFrame> {
+        frame.symbols().iter().map(NoticeFrame::from).collect()
+    }
+}
+
+impl From<&BacktraceSymbol> for NoticeFrame {
+    fn from(symbol: &BacktraceSymbol) -> NoticeFrame {
+        let filename = match symbol.filename() {
+            Some(f) => f
+                .to_str()
+                .expect("Backtrace frame's filename was not valid unicode")
```

Comment:
These expects might be a bit dangerous 🤔At very minimum they'll need extensive tests and documentation about fail cases, but I'd feel much better if they could be removed entirely